{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create SRA submission sheet\n",
    "This Python Jupyter notebook creates a BioSample submission sheet for the Sequence Read Archive:\n",
    "\n",
    "First, import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import glob\n",
    "import itertools\n",
    "import os\n",
    "\n",
    "import natsort\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Display full columns of data frames:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the configuration for the uploads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('upload_config.yaml') as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the Illumina runs to submit:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading Illumina runs from barcode_runs_to_upload.csv\n",
      "Here are the first few entries in the Illumina runs:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>experiment</th>\n",
       "      <th>target</th>\n",
       "      <th>library</th>\n",
       "      <th>antibody</th>\n",
       "      <th>concentration</th>\n",
       "      <th>sort_bin</th>\n",
       "      <th>selection</th>\n",
       "      <th>sample</th>\n",
       "      <th>experiment_type</th>\n",
       "      <th>number_cells</th>\n",
       "      <th>frac_escape</th>\n",
       "      <th>R1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>220422</td>\n",
       "      <td>exptREF</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>lib12</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>ref</td>\n",
       "      <td>reference</td>\n",
       "      <td>exptREF-none-0-ref</td>\n",
       "      <td>ab_selection</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref1_S33_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref2_S34_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref13_S35_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref4_S36_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>220422</td>\n",
       "      <td>exptREF</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>lib13</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>ref</td>\n",
       "      <td>reference</td>\n",
       "      <td>exptREF-none-0-ref</td>\n",
       "      <td>ab_selection</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib2_ref1_S37_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib2_ref2_S38_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib2_ref3_S39_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib2_ref4_S40_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>220422</td>\n",
       "      <td>expt1</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>lib12</td>\n",
       "      <td>LY-CoV1404</td>\n",
       "      <td>154</td>\n",
       "      <td>abneg</td>\n",
       "      <td>escape</td>\n",
       "      <td>expt1-LY-CoV1404-154-abneg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>276693.0</td>\n",
       "      <td>0.075</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/expt1_lib1_S1_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>220422</td>\n",
       "      <td>expt1</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>lib13</td>\n",
       "      <td>LY-CoV1404</td>\n",
       "      <td>154</td>\n",
       "      <td>abneg</td>\n",
       "      <td>escape</td>\n",
       "      <td>expt1-LY-CoV1404-154-abneg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>294136.0</td>\n",
       "      <td>0.075</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/expt1_lib2_S2_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>220422</td>\n",
       "      <td>exptREF</td>\n",
       "      <td>Omicron_BA1</td>\n",
       "      <td>lib22</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>ref</td>\n",
       "      <td>reference</td>\n",
       "      <td>exptREF-none-0-ref</td>\n",
       "      <td>ab_selection</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/220504_VH00319_203_AAAYKKMM5/Unaligned/Project_tstarr/BA1_lib1_ref1_S54_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220504_VH00319_203_AAAYKKMM5/Unaligned/Project_tstarr/BA1_lib1_ref2_S55_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220504_VH00319_203_AAAYKKMM5/Unaligned/Project_tstarr/BA1_lib1_ref13_S56_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220504_VH00319_203_AAAYKKMM5/Unaligned/Project_tstarr/BA1_lib1_ref4_S57_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     date experiment       target library    antibody  concentration sort_bin  \\\n",
       "0  220422    exptREF   Wuhan_Hu_1   lib12        none              0      ref   \n",
       "1  220422    exptREF   Wuhan_Hu_1   lib13        none              0      ref   \n",
       "2  220422      expt1   Wuhan_Hu_1   lib12  LY-CoV1404            154    abneg   \n",
       "3  220422      expt1   Wuhan_Hu_1   lib13  LY-CoV1404            154    abneg   \n",
       "4  220422    exptREF  Omicron_BA1   lib22        none              0      ref   \n",
       "\n",
       "   selection                      sample experiment_type  number_cells  \\\n",
       "0  reference          exptREF-none-0-ref    ab_selection           NaN   \n",
       "1  reference          exptREF-none-0-ref    ab_selection           NaN   \n",
       "2     escape  expt1-LY-CoV1404-154-abneg             NaN      276693.0   \n",
       "3     escape  expt1-LY-CoV1404-154-abneg             NaN      294136.0   \n",
       "4  reference          exptREF-none-0-ref    ab_selection           NaN   \n",
       "\n",
       "   frac_escape  \\\n",
       "0          NaN   \n",
       "1          NaN   \n",
       "2        0.075   \n",
       "3        0.075   \n",
       "4          NaN   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    R1  \n",
       "0  /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref1_S33_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref2_S34_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref13_S35_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref4_S36_R1_001.fastq.gz  \n",
       "1   /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib2_ref1_S37_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib2_ref2_S38_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib2_ref3_S39_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib2_ref4_S40_R1_001.fastq.gz  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                      /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/expt1_lib1_S1_R1_001.fastq.gz  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                      /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/expt1_lib2_S2_R1_001.fastq.gz  \n",
       "4  /shared/ngs/illumina/tstarr/220504_VH00319_203_AAAYKKMM5/Unaligned/Project_tstarr/BA1_lib1_ref1_S54_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220504_VH00319_203_AAAYKKMM5/Unaligned/Project_tstarr/BA1_lib1_ref2_S55_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220504_VH00319_203_AAAYKKMM5/Unaligned/Project_tstarr/BA1_lib1_ref13_S56_R1_001.fastq.gz; /shared/ngs/illumina/tstarr/220504_VH00319_203_AAAYKKMM5/Unaligned/Project_tstarr/BA1_lib1_ref4_S57_R1_001.fastq.gz  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Reading Illumina runs from {config['barcode_runs']}\")\n",
    "\n",
    "illumina_runs = pd.read_csv(config['barcode_runs'])\n",
    "\n",
    "print('Here are the first few entries in the Illumina runs:')\n",
    "illumina_runs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next make submission entries for the barcode runs, initially in \"tidy\" format with one FASTQ file per row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here are the first few submission entries in tidy format:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename_fullpath</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref1_S33_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref1_S33_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref2_S34_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref2_S34_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref13_S35_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref13_S35_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref4_S36_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref4_S36_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib13_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib2_ref1_S37_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib2_ref1_S37_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  biosample_accession                library_ID  \\\n",
       "0        SAMN30603977  lib12_exptREF-none-0-ref   \n",
       "1        SAMN30603977  lib12_exptREF-none-0-ref   \n",
       "2        SAMN30603977  lib12_exptREF-none-0-ref   \n",
       "3        SAMN30603977  lib12_exptREF-none-0-ref   \n",
       "4        SAMN30603977  lib13_exptREF-none-0-ref   \n",
       "\n",
       "                                                                                                         title  \\\n",
       "0  Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "1  Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "2  Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "3  Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "4  Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "\n",
       "  library_strategy library_source library_selection library_layout  platform  \\\n",
       "0         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "1         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "2         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "3         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "4         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "\n",
       "  instrument_model                 design_description filetype  \\\n",
       "0     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "1     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "2     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "3     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "4     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "\n",
       "                                                                                                      filename_fullpath  \\\n",
       "0   /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref1_S33_R1_001.fastq.gz   \n",
       "1   /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref2_S34_R1_001.fastq.gz   \n",
       "2  /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref13_S35_R1_001.fastq.gz   \n",
       "3   /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib1_ref4_S36_R1_001.fastq.gz   \n",
       "4   /shared/ngs/illumina/tstarr/220824_VH00699_190_AAC5MW3M5/Unaligned/Project_tstarr/WH1_lib2_ref1_S37_R1_001.fastq.gz   \n",
       "\n",
       "                             filename  \n",
       "0   WH1_lib1_ref1_S33_R1_001.fastq.gz  \n",
       "1   WH1_lib1_ref2_S34_R1_001.fastq.gz  \n",
       "2  WH1_lib1_ref13_S35_R1_001.fastq.gz  \n",
       "3   WH1_lib1_ref4_S36_R1_001.fastq.gz  \n",
       "4   WH1_lib2_ref1_S37_R1_001.fastq.gz  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submissions_tidy = (\n",
    "    illumina_runs\n",
    "    .assign(\n",
    "        biosample_accession=lambda x: config['biosample_accession'],\n",
    "        library_ID=lambda x: x['library'] + '_' + x['sample'],\n",
    "        title=lambda x: 'Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD ' + x['target'],\n",
    "        library_strategy='AMPLICON',\n",
    "        library_source='SYNTHETIC',\n",
    "        library_selection='PCR',\n",
    "        library_layout='single',\n",
    "        platform='ILLUMINA',\n",
    "        instrument_model='NextSeq 2000',\n",
    "        design_description='PCR of barcodes from RBD variants',\n",
    "        filetype='fastq',\n",
    "        filename_fullpath=lambda x: x['R1'].str.split(';')\n",
    "                                    .map(lambda flist: list(itertools.chain.from_iterable(glob.glob(f.strip()) for f in flist))),       \n",
    "        )\n",
    "    .explode('filename_fullpath')\n",
    "    .assign(filename_fullpath=lambda x: x['filename_fullpath'].str.strip(),\n",
    "            filename=lambda x: x['filename_fullpath'].map(os.path.basename))\n",
    "    .drop(columns=illumina_runs.columns)\n",
    "    .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "assert submissions_tidy['filename_fullpath'].map(os.path.isfile).all()\n",
    "\n",
    "print('Here are the first few submission entries in tidy format:')\n",
    "submissions_tidy.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the actual submission, we need a \"wide\" data frame that for each unique `sample_name` / `library_ID` gives all of the files each in different columns.\n",
    "These should be files without the full path.\n",
    "\n",
    "First, look at how many files there are for each sample / library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>n_files</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib13_exptREF-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib22_exptREF-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib23_exptREF-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib24_exptREF-none-0-ref</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib25_exptREF-none-0-ref</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib12_expt1-LY-CoV1404-154-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib13_expt1-LY-CoV1404-154-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib22_expt25-LY-CoV1404-83-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib23_expt25-LY-CoV1404-83-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib24_expt9-LY-CoV1404-119-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib25_expt9-LY-CoV1404-119-abneg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   biosample_accession                        library_ID  n_files\n",
       "0         SAMN30603977          lib12_exptREF-none-0-ref        4\n",
       "1         SAMN30603977          lib13_exptREF-none-0-ref        4\n",
       "2         SAMN30603977          lib22_exptREF-none-0-ref        4\n",
       "3         SAMN30603977          lib23_exptREF-none-0-ref        4\n",
       "4         SAMN30603977          lib24_exptREF-none-0-ref        4\n",
       "5         SAMN30603977          lib25_exptREF-none-0-ref        2\n",
       "6         SAMN30603977  lib12_expt1-LY-CoV1404-154-abneg        1\n",
       "7         SAMN30603977  lib13_expt1-LY-CoV1404-154-abneg        1\n",
       "8         SAMN30603977  lib22_expt25-LY-CoV1404-83-abneg        1\n",
       "9         SAMN30603977  lib23_expt25-LY-CoV1404-83-abneg        1\n",
       "10        SAMN30603977  lib24_expt9-LY-CoV1404-119-abneg        1\n",
       "11        SAMN30603977  lib25_expt9-LY-CoV1404-119-abneg        1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(submissions_tidy\n",
    " .groupby(['biosample_accession', 'library_ID'])\n",
    " .aggregate(n_files=pd.NamedAgg('filename_fullpath', 'count'))\n",
    " .sort_values('n_files', ascending=False)\n",
    " .reset_index()\n",
    " )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now make the wide submission data frame.\n",
    "Note we keep only the filename column with the path lacking the full directory information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here are the first few submission entries in wide format:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename</th>\n",
       "      <th>filename2</th>\n",
       "      <th>filename3</th>\n",
       "      <th>filename4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib12_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>WH1_lib1_ref1_S33_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref2_S34_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref13_S35_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib1_ref4_S36_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib13_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>WH1_lib2_ref1_S37_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib2_ref2_S38_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib2_ref3_S39_R1_001.fastq.gz</td>\n",
       "      <td>WH1_lib2_ref4_S40_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib12_expt1-LY-CoV1404-154-abneg</td>\n",
       "      <td>Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>expt1_lib1_S1_R1_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib13_expt1-LY-CoV1404-154-abneg</td>\n",
       "      <td>Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>expt1_lib2_S2_R1_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN30603977</td>\n",
       "      <td>lib22_exptREF-none-0-ref</td>\n",
       "      <td>Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Omicron_BA1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 2000</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>BA1_lib1_ref1_S54_R1_001.fastq.gz</td>\n",
       "      <td>BA1_lib1_ref2_S55_R1_001.fastq.gz</td>\n",
       "      <td>BA1_lib1_ref13_S56_R1_001.fastq.gz</td>\n",
       "      <td>BA1_lib1_ref4_S57_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  biosample_accession                        library_ID  \\\n",
       "0        SAMN30603977          lib12_exptREF-none-0-ref   \n",
       "1        SAMN30603977          lib13_exptREF-none-0-ref   \n",
       "2        SAMN30603977  lib12_expt1-LY-CoV1404-154-abneg   \n",
       "3        SAMN30603977  lib13_expt1-LY-CoV1404-154-abneg   \n",
       "4        SAMN30603977          lib22_exptREF-none-0-ref   \n",
       "\n",
       "                                                                                                          title  \\\n",
       "0   Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "1   Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "2   Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "3   Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Wuhan_Hu_1   \n",
       "4  Illumina barcode sequencing from LY-CoV1404 mutational antigenic profiling of the SARS-CoV-2 RBD Omicron_BA1   \n",
       "\n",
       "  library_strategy library_source library_selection library_layout  platform  \\\n",
       "0         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "1         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "2         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "3         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "4         AMPLICON      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "\n",
       "  instrument_model                 design_description filetype  \\\n",
       "0     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "1     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "2     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "3     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "4     NextSeq 2000  PCR of barcodes from RBD variants    fastq   \n",
       "\n",
       "                            filename                          filename2  \\\n",
       "0  WH1_lib1_ref1_S33_R1_001.fastq.gz  WH1_lib1_ref2_S34_R1_001.fastq.gz   \n",
       "1  WH1_lib2_ref1_S37_R1_001.fastq.gz  WH1_lib2_ref2_S38_R1_001.fastq.gz   \n",
       "2      expt1_lib1_S1_R1_001.fastq.gz                                NaN   \n",
       "3      expt1_lib2_S2_R1_001.fastq.gz                                NaN   \n",
       "4  BA1_lib1_ref1_S54_R1_001.fastq.gz  BA1_lib1_ref2_S55_R1_001.fastq.gz   \n",
       "\n",
       "                            filename3                          filename4  \n",
       "0  WH1_lib1_ref13_S35_R1_001.fastq.gz  WH1_lib1_ref4_S36_R1_001.fastq.gz  \n",
       "1   WH1_lib2_ref3_S39_R1_001.fastq.gz  WH1_lib2_ref4_S40_R1_001.fastq.gz  \n",
       "2                                 NaN                                NaN  \n",
       "3                                 NaN                                NaN  \n",
       "4  BA1_lib1_ref13_S56_R1_001.fastq.gz  BA1_lib1_ref4_S57_R1_001.fastq.gz  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submissions_wide = (\n",
    "    submissions_tidy\n",
    "    .assign(\n",
    "        filename_count=lambda x: x.groupby(['biosample_accession', 'library_ID'])['filename'].cumcount() + 1,\n",
    "        filename_col=lambda x: 'filename' + x['filename_count'].map(lambda c: str(c) if c > 1 else '')\n",
    "        )\n",
    "    .pivot(\n",
    "        index='library_ID',\n",
    "        columns='filename_col',\n",
    "        values='filename',\n",
    "        )\n",
    "    )\n",
    "\n",
    "submissions_wide = (\n",
    "    submissions_tidy\n",
    "    .drop(columns=['filename_fullpath', 'filename'])\n",
    "    .drop_duplicates()\n",
    "    .merge(submissions_wide[natsort.natsorted(submissions_wide.columns)],\n",
    "           on='library_ID',\n",
    "           validate='one_to_one',\n",
    "           )\n",
    "    )\n",
    "\n",
    "print('Here are the first few submission entries in wide format:')\n",
    "submissions_wide.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now write the wide submissions data frame to a `*.tsv` file that can be used for uploading the SRA submission website:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "submissions_spreadsheet = 'SRA_submission_spreadsheet.tsv'\n",
    "\n",
    "submissions_wide.to_csv(submissions_spreadsheet, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also want to write a file with all of the FASTQ files that are in the submission spreadsheet:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing list of all the FASTQ files to upload to FASTQs_to_upload.csv\n"
     ]
    }
   ],
   "source": [
    "fastq_file_list = 'FASTQs_to_upload.csv'\n",
    "\n",
    "print(f\"Writing list of all the FASTQ files to upload to {fastq_file_list}\")\n",
    "\n",
    "(submissions_tidy\n",
    " [['filename_fullpath', 'filename']]\n",
    " .to_csv(fastq_file_list, index=False)\n",
    " )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
